### ISSUE TITLE
v6 source-loader / addon-storysource not working, when default and named export of stories are defined in two different files

### ISSUE THREAD

**Issue by vinothvk** [storybookjs/storybook#12199]
**Describe the bug**
When we separate the default export of a story and the definition of a named exported story into two different files, then source loader / addon-storysource modules are not working. Story tab in tha add-on panel shows up as empty

This scenario works fine in v5 but fails in v6

**To Reproduce**
download the sample reproducible code from below repo and run npm run storybook
https://github.com/vinothvk/storybookjv6-story-source-issue

Now visit 
http://localhost:9010/?path=/story/greeting--title-only 
http://localhost:9010/?path=/story/greeting--with-message

check the story tab it will be empty

**Expected behavior**
story definition though it is present in different file should show up (it does in v5 version not in v6)
Also the line story start and end to be highlighted ( this was not working in v5 )

**Screenshots**
NA

**Code snippets**
https://github.com/vinothvk/storybookjv6-story-source-issue

**System:**
Environment Info:

  System:
    OS: macOS Mojave 10.14.6
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  Binaries:
    Node: 12.18.3 - ~/.nvm/versions/node/v12.18.3/bin/node
    Yarn: 1.22.4 - ~/.yarn/bin/yarn
    npm: 6.14.7 - ~/.nvm/versions/node/v12.18.3/bin/npm
  Browsers:
    Chrome: 84.0.4147.135
    Firefox: 78.0.2
    Safari: 13.1.2
  npmPackages:
    @storybook/addon-a11y: ~6.0.16 => 6.0.16
    @storybook/addon-actions: ~6.0.16 => 6.0.16
    @storybook/addon-backgrounds: ~6.0.16 => 6.0.16
    @storybook/addon-console: ~1.2.1 => 1.2.1
    @storybook/addon-controls: ~6.0.16 => 6.0.16
    @storybook/addon-docs: ~6.0.16 => 6.0.16
    @storybook/addon-events: ~6.0.16 => 6.0.16
    @storybook/addon-knobs: ~6.0.16 => 6.0.16
    @storybook/addon-links: ~6.0.16 => 6.0.16
    @storybook/addon-storysource: ~6.0.16 => 6.0.16
    @storybook/addon-toolbars: ~6.0.16 => 6.0.16
    @storybook/addon-viewport: ~6.0.16 => 6.0.16
    @storybook/addons: ~6.0.16 => 6.0.16
    @storybook/react: ~6.0.16 => 6.0.16
    @storybook/theming: ~6.0.16 => 6.0.16

**Additional context**
NA

**Comment by vinothvk** [comment]
@shilman i am sure you and team are busy in tackling things. Just want to know whether this issue has any priority / timeline in bugs to be resolved list ?

**Comment by shilman** [comment]
which version are you upgrading from? we're not prioritizing storysource bugs in general (and are looking for maintainers) but source-loader is currently used in addon-docs, so that aspect could be prioritized

**Comment by vinothvk** [comment]
@shilman we upgraded from v5 to v6. I attached the link to the source code that reproduces the issue.
The issue when analyzed seems to be in source-loader than addon-storysource. if you can prioritize it that would be great.

**Comment by hbmuller** [comment]
Facing a similar problem over here. We're trying to adopt storysource, but some of our component require nesting many subcomponents into them. Eg.: `Table`, which requires head, body, rows, cells and other related components to be deeply nested into it.

That causes each named export to be a long component. And having multiple of them in a single file is exactly what we are trying to avoid. So we broke them down into multiple files within a story folder.

---

**Folder structure:**
- .storybook/
  - main.js
- src/
  - simple.stories.tsx (this works - has both default and named exports)
  - complex.stories/
    - index.ts
    - example-a.tsx

```js
// .storybook/main.js

module.exports = {
  // ...
  stories: [
    "../src/stories/**/*.stories.@(ts|tsx)",
    "../src/stories/**/*.stories/index.@(ts|tsx)",
  ],
  // ...
};
```

```ts
// complex.stories/index.ts

import { Meta } from "@storybook/react/types-6-0";
import { Table } from "component-library";

export default {
  title: "Table",
  component: Table,
} as Meta;

export * from "./example-a";
// export * from "./example-b";
// export * from "./example-c";
```

```tsx
// complex.stories/example-a.tsx

import * as React from "react";
import { Story } from "@storybook/react/types-6-0";
import {
  Table,
  // ...
} from "component-library";

export const ExampleA: Story = () => (
  <Table>
    {/* Complex table example */}
  </Table>
);
```

---

For `simple.stories.tsx` we get the source code in the tab as expected. But for `complex.stories/example-a.tsx` the tab is blank.

**Comment by vinothvk** [comment]
@shilman do you have any timeline / roadmap in mind for the fix?
thanks

**Comment by shilman** [comment]
No, we'll working through a backlog of issues for essential addons and that's getting our priority right now https://storybook.js.org/docs/react/essentials/introduction. I would happily review a PR with the fix.

**Comment by nickofthyme** [comment]
Ok after scouring through the `src` code and much trial and error, I have found a temporary workaround for `^6.3.7`(haven't tested other versions). I have a slightly different setup than the one defined above but given the following example setup below and steps, I think we can resolve any setup.

```
.storybook/
├── main.js
├── preview.js
└── stories
    ├── complex_story.tsx
    └── my_stories.stories.tsx
```

## Required changes

### Rename all complex stories to `*.story.tsx`

The [`@storybook/source-loader`](https://www.npmjs.com/package/@storybook/source-loader), which is in fact a custom webpack loader, is used here...

https://github.com/storybookjs/storybook/blob/337fdcd0fe7436b46bcca65145ff6db2affd780f/addons/docs/src/frameworks/common/preset.ts#L54-L63

Since the `test` path above only includes `*.story.*` or `*.stories.*` files, our `complex_story.tsx` file will be excluded and thus the story source will never find the file. So our setup now looks like this...

```diff
.storybook/
├── main.js
├── preview.js
└── stories
-   ├── complex_story.tsx
+   ├── complex_story.story.tsx
    └── my_stories.stories.tsx
```

> Note: you could alternatively override the default webpack rule and set the `test` regex to include your complex story files but I find this to be easier.

### Include your story files in the `main.stories` array

The source will not be discoverable unless the files are included in the `stories` path. Thus, `complex_story.story.tsx` __must__ be included.

```diff
// main.js
module.exports = {
  stories: [
-   './stories/**/*.stories.ts?(x)",
+   './stories/**/*.(stories|story).ts?(x)",
  ],
  // ...
};
```

### Include `@storybook/addon-docs`

For whatever reason, you must also include [`@storybook/addon-docs`](https://storybook.js.org/addons/@storybook/addon-docs), without this addon the source-loader/story-source still cannot find the correct export to use as source. This must contain some logic for connecting exports to their original file.

```diff
// main.js
module.exports = {
  addons: [
    '@storybook/addon-storysource'
+   '@storybook/addon-docs'
  ],
  // ...
};
```

If you, like me, are not using the docs tab you can easily hide it like so...

```js
// preview.js
export const parameters  = {
  previewTabs: {
    'storybook/docs/panel': {
      hidden: true,
    },
  },
};
```

## Caveats

- You cannot use `export default` from these `complex_story.story.tsx` files otherwise they will be treated as stories and require the proper `title` etc., and likely be a duplicate story.
- This workaround will only show the source of the exported value and not the file from which it was exported. So `import`s and anything else outside of the specific export with be excluded.

## Conclusion

I do not think this is a solution to the issue as it requires non-obvious implementation to fix. Also the functionality of prior versions (`^5.x.x`) allowed for the full file to be visible and not just the exported content.

See https://github.com/nickofthyme/storybook/commit/d7429cf3592f67eaf992525f90e07f9ccb70f301 for working demo, just run `yarn bootstrap --core && yarn start` to run locally.

**Comment by nickofthyme** [comment]
Follow up I am noticing an issue similar to https://github.com/storybookjs/storybook/issues/12208 in some stories.

**Comment by vanessayuenn** [comment]
`addon-storysource` has been deprecated. As of Storybook 9, it has been replaced by the [addon-docs code panel](https://storybook.js.org/docs/writing-docs/code-panel), which provides more snippet options. Please see [migration guide](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#storysource-addon-removed).
